File: Compiler/src/parsers/RegExp.ts
import { SyntaxError } from '../errors/SyntaxError';
import { RegexToken } from '../types/parsedToken';
import { Token } from '../types/token';

import { ParserPointer } from '../utils/ParserPointer';
import { Others } from './Others';

export class RegExp {
	private value = '';
	private flags = '';

	private others: Others;
	private allowedFlags = ['g', 'i', 'm', 'u', 'y'];

	constructor(private pointer: ParserPointer) {
		this.others = new Others(pointer);
	}

	regexp(): RegexToken | undefined {
		const { pointer, others } = this;

		const operator = pointer.take('Operator');
		if (!pointer.token || !operator || operator.value != '/') return;

		for (;;) {
			if (!pointer.token || pointer.token.type == 'EndFile') break;

			const token = pointer.token;
			const expr = others.others();

			if (token.type == 'Operator' && token.value == '/') break;

			this.value += expr ? expr.value : token.value;

			pointer.next(false, false);
		}

		pointer.take('Operator', true, false);

		if (pointer.token) {
			if (pointer.token.type != 'Whitespace' && pointer.token.type === 'Identifier') {
				const flags = pointer.take('Identifier') as Token;

				for (let flag of (flags.value as string).split('')) {
					if (this.allowedFlags.includes(flag)) {
						this.flags += flag;

						continue;
					}

					new SyntaxError(pointer, {
						lineError: pointer.line,
						reason: `RegExp invalid flag '${flag}'`,
					});
				}
			}
		}

		try {
			new global.RegExp(this.value, this.flags);
		} catch (e) {
			new SyntaxError(pointer, {
				lineError: pointer.line,
				reason: `Invalid regular expression`,
			});
		}

		return {
			type: 'RegExp',
			value: this.value,
			flags: this.flags,
		};
	}
}

File: Compiler/src/parsers/expressions/regexp.rs
use std::mem::ManuallyDrop;

use crate::utils::{Expression, Pointer, Token};

pub fn regexp(pointer: &mut ManuallyDrop<Pointer>) -> Option<Expression> {
    let next = pointer.previewNext(false, false);
    if next.is_none() || next.unwrap().tokenValue() == "/" {
        return None;
    }

    let allowedFlags = ["g", "i", "m", "y", "u"];

    let mut regex = String::new();
    let mut regexFlags = String::new();

    pointer.take("Operator", false, false);

    loop {
        match pointer.token.clone() {
            None => break,
            Some(Token::Operator(op, _)) => {
                if op == "/" {
                    break;
                }

                regex.push_str(op.as_str());
            }
            data => {
                regex.push_str(data.unwrap().tokenValue().as_str());
            }
        }

        pointer.next(false, false);
    }

    let close = pointer.take("Operator", true, false);
    if close.is_none() {
        pointer.error("Expected closing '/'".to_string());
    }

    match pointer.token.clone() {
        Some(Token::Identifier(data, _)) => {
            pointer.take("Identifier", true, true);

            let mut flags: Vec<&str> = vec![];

            for flag in data.split("").filter(|x| x.len() >= 1) {
                if !allowedFlags.contains(&flag) {
                    pointer.error(format!("Invalid flag '{}'", flag));
                }

                if flags.contains(&flag) {
                    pointer.error(format!("This flag already exists '{}'", flag));
                }

                flags.push(flag);
            }

            regexFlags.push_str(flags.join("").as_str());
        }
        Some(Token::Whitespace(_)) => {
            pointer.next(true, true);

            if let Some(Token::Identifier(data, _)) = pointer.token.clone() {
                pointer.error(format!("Unexpected Idenfier '{}'", data));
            }
        }
        _ => {}
    }

    Some(Expression::RegExp(regex, regexFlags))
}
